File: java/jscore-jvm/src/test/kotlin/com/o2ter/jscore/network/WebSocketTests.kt
//
//  WebSocketTests.kt
//
//  The MIT License
//  Copyright (c) 2021 - 2025 O2ter Limited. All rights reserved.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.
//

package com.o2ter.jscore.network

import com.o2ter.jscore.JavaScriptEngine
import com.o2ter.jscore.jvm.JvmPlatformContext
import com.o2ter.jscore.executeAsync
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class WebSocketTests {
    
    @Test
    fun testWebSocketExists() {
        val context = JvmPlatformContext("TestApp")
        val engine = JavaScriptEngine(context)
        try {
            val result = engine.execute("typeof WebSocket")
            assertEquals("function", result.toString())
        } finally {
            engine.close()
        }
    }
    
    @Test
    fun testWebSocketConstants() {
        val context = JvmPlatformContext("TestApp")
        val engine = JavaScriptEngine(context)
        try {
            val result = engine.execute("""
                ({
                    connecting: WebSocket.CONNECTING,
                    open: WebSocket.OPEN,
                    closing: WebSocket.CLOSING,
                    closed: WebSocket.CLOSED
                })
            """.trimIndent()) as? Map<*, *>
            
            assertNotNull(result)
            val connecting = (result!!["connecting"] as Number).toDouble()
            val open = (result["open"] as Number).toDouble()
            val closing = (result["closing"] as Number).toDouble()
            val closed = (result["closed"] as Number).toDouble()
            
            assertEquals(0.0, connecting)
            assertEquals(1.0, open)
            assertEquals(2.0, closing)
            assertEquals(3.0, closed)
        } finally {
            engine.close()
        }
    }
    
    @Test
    fun testWebSocketInstantiation() {
        val context = JvmPlatformContext("TestApp")
        val engine = JavaScriptEngine(context)
        try {
            val result = executeAsync(engine, """
                (async () => {
                    return new Promise((resolve, reject) => {
                        try {
                            const ws = new WebSocket('wss://echo.websocket.org');
                            const result = {
                                hasUrl: typeof ws.url === 'string',
                                hasReadyState: typeof ws.readyState === 'number',
                                hasBufferedAmount: typeof ws.bufferedAmount === 'number',
                                hasSend: typeof ws.send === 'function',
                                hasClose: typeof ws.close === 'function',
                                initialState: ws.readyState
                            };
                            ws.close();
                            resolve(result);
                        } catch (error) {
                            reject(error);
                        }
                    });
                })()
            """.trimIndent(), timeoutMs = 10000) as? Map<*, *>
            
            assertNotNull(result)
            assertEquals(true, result!!["hasUrl"])
            assertEquals(true, result["hasReadyState"])
            assertEquals(true, result["hasBufferedAmount"])
            assertEquals(true, result["hasSend"])
            assertEquals(true, result["hasClose"])
            assertEquals(0.0, (result["initialState"] as Number).toDouble())
        } finally {
            engine.close()
        }
    }
    
    @Test
    fun testWebSocketConnection() {
        val context = JvmPlatformContext("TestApp")
        val engine = JavaScriptEngine(context)
        try {
            val result = executeAsync(engine, """
                (async () => {
                    return new Promise((resolve, reject) => {
                        const ws = new WebSocket('wss://echo.websocket.org');
                        let openFired = false;
                        
                        ws.onopen = function() {
                            openFired = true;
                            resolve({
                                opened: true,
                                readyState: ws.readyState
                            });
                            ws.close();
                        };
                        
                        ws.onerror = function(event) {
                            reject(new Error('Connection error'));
                        };
                        
                        setTimeout(() => {
                            if (!openFired) {
                                ws.close();
                                reject(new Error('Connection timeout'));
                            }
                        }, 10000);
                    });
                })()
            """.trimIndent(), timeoutMs = 15000) as? Map<*, *>
            
            assertNotNull(result)
            assertEquals(true, result!!["opened"])
            assertEquals(1.0, (result["readyState"] as Number).toDouble())
        } finally {
            engine.close()
        }
    }
    
    @Test
    fun testWebSocketSendReceiveText() {
        val context = JvmPlatformContext("TestApp")
        val engine = JavaScriptEngine(context)
        try {
            val result = executeAsync(engine, """
                (async () => {
                    return new Promise((resolve, reject) => {
                        const ws = new WebSocket('wss://echo.websocket.org');
                        const testMessage = 'Hello WebSocket!';
                        let messageReceived = false;
                        
                        ws.onopen = function() {
                            ws.send(testMessage);
                        };
                        
                        ws.onmessage = function(event) {
                            messageReceived = true;
                            resolve({
                                success: true,
                                messageReceived: typeof event.data === 'string'
                            });
                            ws.close();
                        };
                        
                        ws.onerror = function() {
                            reject(new Error('Connection error'));
                        };
                        
                        setTimeout(() => {
                            if (!messageReceived) {
                                ws.close();
                                reject(new Error('Message timeout'));
                            }
                        }, 10000);
                    });
                })()
            """.trimIndent(), timeoutMs = 15000) as? Map<*, *>
            
            assertNotNull(result)
            assertEquals(true, result!!["success"])
            assertEquals(true, result["messageReceived"])
        } finally {
            engine.close()
        }
    }
    
    @Test
    fun testWebSocketClose() {
        val context = JvmPlatformContext("TestApp")
        val engine = JavaScriptEngine(context)
        try {
            val result = executeAsync(engine, """
                (async () => {
                    return new Promise((resolve, reject) => {
                        const ws = new WebSocket('wss://echo.websocket.org');
                        let closeFired = false;
                        
                        ws.onopen = function() {
                            ws.close(1000, 'Normal closure');
                        };
                        
                        ws.onclose = function(event) {
                            closeFired = true;
                            resolve({
                                closed: true,
                                code: event.code,
                                reason: event.reason,
                                wasClean: event.wasClean,
                                readyState: ws.readyState
                            });
                        };
                        
                        ws.onerror = function() {
                            reject(new Error('Connection error'));
                        };
                        
                        setTimeout(() => {
                            if (!closeFired) {
                                reject(new Error('Close timeout'));
                            }
                        }, 10000);
                    });
                })()
            """.trimIndent(), timeoutMs = 15000) as? Map<*, *>
            
            assertNotNull(result)
            assertEquals(true, result!!["closed"])
            assertEquals(1000.0, (result["code"] as Number).toDouble())
            assertEquals(true, result["wasClean"])
            assertEquals(3.0, (result["readyState"] as Number).toDouble())
        } finally {
            engine.close()
        }
    }
    
    @Test
    fun testWebSocketBinaryType() {
        val context = JvmPlatformContext("TestApp")
        val engine = JavaScriptEngine(context)
        try {
            val result = engine.execute("""
                const ws = new WebSocket('wss://echo.websocket.org');
                const initialType = ws.binaryType;
                ws.binaryType = 'arraybuffer';
                const newType = ws.binaryType;
                ws.close();
                
                ({
                    initialType: initialType,
                    newType: newType
                })
            """.trimIndent()) as? Map<*, *>
            
            assertNotNull(result)
            assertEquals("blob", result!!["initialType"])
            assertEquals("arraybuffer", result["newType"])
        } finally {
            engine.close()
        }
    }
    
    @Test
    fun testWebSocketInvalidURL() {
        val context = JvmPlatformContext("TestApp")
        val engine = JavaScriptEngine(context)
        try {
            // Invalid protocol should be rejected by polyfill validation
            val script = """
                let errorCaught = false;
                try {
                    new WebSocket('http://invalid-protocol.com');
                } catch (e) {
                    errorCaught = e.message.includes('invalid');
                }
                errorCaught;
            """.trimIndent()
            val result = engine.execute(script)
            assertEquals(true, result, "Should catch error for invalid protocol")
        } finally {
            engine.close()
        }
    }
    
    @Test
    fun testWebSocketEventListeners() {
        val context = JvmPlatformContext("TestApp")
        val engine = JavaScriptEngine(context)
        try {
            val result = executeAsync(engine, """
                (async () => {
                    return new Promise((resolve, reject) => {
                        const ws = new WebSocket('wss://echo.websocket.org');
                        let openCount = 0;
                        
                        ws.addEventListener('open', function() {
                            openCount++;
                            if (openCount === 1) {
                                resolve({
                                    eventListenerWorks: true,
                                    openCount: openCount
                                });
                                ws.close();
                            }
                        });
                        
                        ws.onerror = function() {
                            reject(new Error('Connection error'));
                        };
                        
                        setTimeout(() => {
                            if (openCount === 0) {
                                ws.close();
                                reject(new Error('Open timeout'));
                            }
                        }, 10000);
                    });
                })()
            """.trimIndent(), timeoutMs = 15000) as? Map<*, *>
            
            assertNotNull(result)
            assertEquals(true, result!!["eventListenerWorks"])
            assertEquals(1.0, (result["openCount"] as Number).toDouble())
        } finally {
            engine.close()
        }
    }
}

File: swift/Tests/SwiftJSTests/Networking/WebSocketTests.swift
//
//  WebSocketTests.swift
//
//  The MIT License
//  Copyright (c) 2021 - 2025 O2ter Limited. All rights reserved.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.
//

import XCTest
@testable import SwiftJS

/// Tests for the WebSocket API implementation
@MainActor
final class WebSocketTests: XCTestCase {
    
    // MARK: - WebSocket API Tests
    
    func testWebSocketExists() {
        let context = SwiftJS()
        let result = context.evaluateScript("typeof WebSocket")
        XCTAssertEqual(result.toString(), "function")
    }
    
    func testWebSocketConstants() {
        let script = """
            ({
                connecting: WebSocket.CONNECTING,
                open: WebSocket.OPEN,
                closing: WebSocket.CLOSING,
                closed: WebSocket.CLOSED
            })
        """
        let context = SwiftJS()
        let result = context.evaluateScript(script)
        
        XCTAssertEqual(Int(result["connecting"].numberValue ?? -1), 0)
        XCTAssertEqual(Int(result["open"].numberValue ?? -1), 1)
        XCTAssertEqual(Int(result["closing"].numberValue ?? -1), 2)
        XCTAssertEqual(Int(result["closed"].numberValue ?? -1), 3)
    }
    
    func testWebSocketInstantiation() {
        let expectation = XCTestExpectation(description: "WebSocket instantiation")
        
        let script = """
            try {
                const ws = new WebSocket('wss://echo.websocket.org');
                const result = {
                    hasUrl: typeof ws.url === 'string',
                    hasReadyState: typeof ws.readyState === 'number',
                    hasBufferedAmount: typeof ws.bufferedAmount === 'number',
                    hasSend: typeof ws.send === 'function',
                    hasClose: typeof ws.close === 'function',
                    initialState: ws.readyState
                };
                ws.close();
                testCompleted(result);
            } catch (error) {
                testCompleted({ error: error.message });
            }
        """
        
        let context = SwiftJS()
        context.globalObject["testCompleted"] = SwiftJS.Value(in: context) { args, this in
            let result = args[0]
            if result["error"].isString {
                XCTFail("Error creating WebSocket: \(result["error"].toString())")
            } else {
                XCTAssertTrue(result["hasUrl"].boolValue ?? false, "Should have url property")
                XCTAssertTrue(result["hasReadyState"].boolValue ?? false, "Should have readyState property")
                XCTAssertTrue(result["hasBufferedAmount"].boolValue ?? false, "Should have bufferedAmount property")
                XCTAssertTrue(result["hasSend"].boolValue ?? false, "Should have send method")
                XCTAssertTrue(result["hasClose"].boolValue ?? false, "Should have close method")
                XCTAssertEqual(Int(result["initialState"].numberValue ?? -1), 0, "Initial state should be CONNECTING")
            }
            expectation.fulfill()
            return SwiftJS.Value.undefined
        }
        
        context.evaluateScript(script)
        wait(for: [expectation], timeout: 10.0)
    }
    
    func testWebSocketConnection() {
        let expectation = XCTestExpectation(description: "WebSocket connection")
        
        let script = """
            const ws = new WebSocket('wss://echo.websocket.org');
            let openFired = false;
            
            ws.onopen = function() {
                openFired = true;
                testCompleted({
                    opened: true,
                    readyState: ws.readyState
                });
                ws.close();
            };
            
            ws.onerror = function(event) {
                testCompleted({ error: 'Connection error' });
            };
            
            // Timeout after 10 seconds
            setTimeout(() => {
                if (!openFired) {
                    ws.close();
                    testCompleted({ timeout: true });
                }
            }, 10000);
        """
        
        let context = SwiftJS()
        context.globalObject["testCompleted"] = SwiftJS.Value(in: context) { args, this in
            let result = args[0]
            if result["error"].isString {
                XCTAssertTrue(true, "Network test skipped: \(result["error"].toString())")
            } else if result["timeout"].boolValue == true {
                XCTAssertTrue(true, "Connection timeout - network might not be available")
            } else {
                XCTAssertTrue(result["opened"].boolValue ?? false, "WebSocket should open")
                XCTAssertEqual(Int(result["readyState"].numberValue ?? -1), 1, "Ready state should be OPEN")
            }
            expectation.fulfill()
            return SwiftJS.Value.undefined
        }
        
        context.evaluateScript(script)
        wait(for: [expectation], timeout: 15.0)
    }
    
    func testWebSocketSendReceiveText() {
        let expectation = XCTestExpectation(description: "WebSocket send/receive text")
        
        let script = """
            const ws = new WebSocket('wss://echo.websocket.org');
            const testMessage = 'Hello WebSocket!';
            let messageReceived = false;
            
            ws.onopen = function() {
                ws.send(testMessage);
            };
            
            ws.onmessage = function(event) {
                messageReceived = true;
                testCompleted({
                    success: true,
                    sentMessage: testMessage,
                    receivedMessage: event.data,
                    messagesMatch: event.data === testMessage
                });
                ws.close();
            };
            
            ws.onerror = function() {
                testCompleted({ error: 'Connection error' });
            };
            
            setTimeout(() => {
                if (!messageReceived) {
                    ws.close();
                    testCompleted({ timeout: true });
                }
            }, 10000);
        """
        
        let context = SwiftJS()
        context.globalObject["testCompleted"] = SwiftJS.Value(in: context) { args, this in
            let result = args[0]
            if result["error"].isString {
                XCTAssertTrue(true, "Network test skipped: \(result["error"].toString())")
            } else if result["timeout"].boolValue == true {
                XCTAssertTrue(true, "Test timeout - network might not be available")
            } else {
                XCTAssertTrue(result["success"].boolValue ?? false, "Should successfully send/receive")
                // Note: echo.websocket.org may return server info instead of echoing, so just verify we got a response
                let received = result["receivedMessage"].toString()
                XCTAssertFalse(received.isEmpty, "Should receive a message (echo server may return server info)")
            }
            expectation.fulfill()
            return SwiftJS.Value.undefined
        }
        
        context.evaluateScript(script)
        wait(for: [expectation], timeout: 15.0)
    }
    
    func testWebSocketClose() {
        let expectation = XCTestExpectation(description: "WebSocket close")
        
        let script = """
            const ws = new WebSocket('wss://echo.websocket.org');
            let closeFired = false;
            
            ws.onopen = function() {
                ws.close(1000, 'Normal closure');
            };
            
            ws.onclose = function(event) {
                closeFired = true;
                testCompleted({
                    closed: true,
                    code: event.code,
                    reason: event.reason,
                    wasClean: event.wasClean,
                    readyState: ws.readyState
                });
            };
            
            ws.onerror = function() {
                testCompleted({ error: 'Connection error' });
            };
            
            setTimeout(() => {
                if (!closeFired) {
                    testCompleted({ timeout: true });
                }
            }, 10000);
        """
        
        let context = SwiftJS()
        context.globalObject["testCompleted"] = SwiftJS.Value(in: context) { args, this in
            let result = args[0]
            if result["error"].isString {
                XCTAssertTrue(true, "Network test skipped: \(result["error"].toString())")
            } else if result["timeout"].boolValue == true {
                XCTAssertTrue(true, "Test timeout - network might not be available")
            } else {
                XCTAssertTrue(result["closed"].boolValue ?? false, "Close event should fire")
                XCTAssertEqual(Int(result["code"].numberValue ?? -1), 1000, "Close code should be 1000")
                XCTAssertTrue(result["wasClean"].boolValue ?? false, "Should be clean closure")
                XCTAssertEqual(Int(result["readyState"].numberValue ?? -1), 3, "Ready state should be CLOSED")
            }
            expectation.fulfill()
            return SwiftJS.Value.undefined
        }
        
        context.evaluateScript(script)
        wait(for: [expectation], timeout: 15.0)
    }
    
    func testWebSocketBinaryType() {
        let script = """
            const ws = new WebSocket('wss://echo.websocket.org');
            const initialType = ws.binaryType;
            ws.binaryType = 'arraybuffer';
            const newType = ws.binaryType;
            ws.close();
            
            ({
                initialType: initialType,
                newType: newType
            })
        """
        
        let context = SwiftJS()
        let result = context.evaluateScript(script)
        
        XCTAssertEqual(result["initialType"].toString(), "blob", "Default binaryType should be blob")
        XCTAssertEqual(result["newType"].toString(), "arraybuffer", "binaryType should change to arraybuffer")
    }
    
    func testWebSocketInvalidURL() {
        let script = """
            try {
                const ws = new WebSocket('http://invalid-protocol.com');
                ({ error: false })
            } catch (error) {
                ({ error: true, message: error.message })
            }
        """
        
        let context = SwiftJS()
        let result = context.evaluateScript(script)
        
        XCTAssertTrue(result["error"].boolValue ?? false, "Should throw error for invalid protocol")
    }
    
    func testWebSocketEventListeners() {
        let expectation = XCTestExpectation(description: "WebSocket event listeners")
        
        let script = """
            const ws = new WebSocket('wss://echo.websocket.org');
            let openCount = 0;
            
            // Test addEventListener
            ws.addEventListener('open', function() {
                openCount++;
                if (openCount === 1) {
                    testCompleted({
                        eventListenerWorks: true,
                        openCount: openCount
                    });
                    ws.close();
                }
            });
            
            ws.onerror = function() {
                testCompleted({ error: 'Connection error' });
            };
            
            setTimeout(() => {
                if (openCount === 0) {
                    ws.close();
                    testCompleted({ timeout: true });
                }
            }, 10000);
        """
        
        let context = SwiftJS()
        context.globalObject["testCompleted"] = SwiftJS.Value(in: context) { args, this in
            let result = args[0]
            if result["error"].isString {
                XCTAssertTrue(true, "Network test skipped: \(result["error"].toString())")
            } else if result["timeout"].boolValue == true {
                XCTAssertTrue(true, "Test timeout - network might not be available")
            } else {
                XCTAssertTrue(result["eventListenerWorks"].boolValue ?? false, "addEventListener should work")
                XCTAssertEqual(Int(result["openCount"].numberValue ?? 0), 1, "Open event should fire once")
            }
            expectation.fulfill()
            return SwiftJS.Value.undefined
        }
        
        context.evaluateScript(script)
        wait(for: [expectation], timeout: 15.0)
    }
}
